# Case Study 4.1 - Movies

<h1 style="color:red;">Note: If you close this notebook at any time, you will have to run all cells again upon re-opening it.</h1>

<h1 style="color:red;">Note: You may get different numerical results running the notebook different times. This is to be expected, you can just report whatever results you get.</h1>

# ADVANCED PYTHON

As this is an advanced version, we don't include a lot of code here. If you get stuck on a particular part, feel free to also use the beginner version in `beginnner_python.ipynb` to help you out.

# Identification Information

In [1]:
# YOUR NAME              = Armen Khachatrian
# YOUR MITX PRO USERNAME = armen-khachatrian
# YOUR MITX PRO E-MAIL   = aa.khachatryan97@gmail.com

# Setup

Run these cells to install all the packages you need to complete the remainder of the case study. This may take a few minutes, so please be patient.

In [2]:
!pip install surprise==0.1

  Using cached https://files.pythonhosted.org/packages/61/de/e5cba8682201fcf9c3719a6fdda95693468ed061945493dea2dd37c5618b/surprise-0.1-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/f5/da/b5700d96495fb4f092be497f02492768a3d96a3f4fa2ae7dea46d4081cfa/scikit-surprise-1.1.0.tar.gz
  ERROR: Command errored out with exit status 1:
   command: /Users/a.khachatryan/anaconda3/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/q7/1xn5m1555ws7kn30v5f1yvyw0000gn/T/pip-install-69frhtmb/scikit-surprise/setup.py'"'"'; __file__='"'"'/private/var/folders/q7/1xn5m1555ws7kn30v5f1yvyw0000gn/T/pip-install-69frhtmb/scikit-surprise/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/q7/1xn5m1555ws7kn30v5f1yvyw0000gn/T/pip-wheel-xwt92ewj --python-tag cp36
       cwd: /private/var/fold

Now, you must press **Kernel > Restart.** This allows the installation to take effect. Once you see the blue **Connected/Kernel ready** button in the top right, you are good to go.

# Import

In [1]:
import pandas as pd
import matplotlib
from surprise import Dataset, SVD, NormalPredictor, BaselineOnly, KNNBasic, NMF
from surprise.model_selection import cross_validate, KFold

ModuleNotFoundError: No module named 'surprise'

# Data

Use the [`**Dataset.load_builtin**`](http://surprise.readthedocs.io/en/stable/dataset.html#surprise.dataset.Dataset.load_builtin) function to load the data.

In [7]:
data = Dataset.load_builtin('ml-100k')

NameError: name 'Dataset' is not defined

We also want to get a sense of what the data looks like. Please create a histogram of all the ratings we have in the dataset.

In [7]:
# 1. Get the ratings file from the data object
# This is just a filename that has all the data stored in it
ratings_file = data.ratings_file


# 2. Load that table using pandas, a commmon python data loading tool
# We set the column names manually here
col_names = ['user_id', 'item_id', 'rating', 'timestamp']
raw_data = pd.read_table(ratings_file, names=col_names)

# 3. Get the rating column
ratings = raw_data.rating

# 4. Generate a bar plot/histogram of that data
ratings.value_counts().sort_index().plot.bar()

In [8]:
len(ratings)

100000

In [9]:
import statistics
mean = statistics.mean(ratings)
median = statistics.median(ratings)
mode = statistics.mode(ratings)
print("Number of observations: ", len(ratings))
print("Mean = ", mean)
print("Median = ", median)
print("Mode = ", mode)


Number of observations:  100000
Mean =  3.52986
Median =  4.0
Mode =  4


<h1 style="color:red;">QUESTION 1: DATA ANALYSIS</h1>

**Describe the dataset. How many ratings are in the dataset? How would you describe the distribution of ratings? Is there anything else we should observe? Make sure the histogram is visible in the notebook.**

#### 1. How many ratings are in the dataset?
100000
#### 2. How would you describe the distribution of ratings?

2.1. **Shape**: approximately symmetric distribution

2.2. **Center**: 
            
            Mean = 3.52986
            
            Median = 4.0
            
            Mode = 4
            
2.3. **Spread**: range = 4

2.4. **Outliers** = No apparent outliers

# Model 1: Random

In [10]:
# Create model using NormalPredictor() class
model_random = NormalPredictor()

In [11]:
# Train on data using cross-validation with k=5 folds, measuring the RMSE
# See the cross_validate function that we have imported above
# http://surprise.readthedocs.io/en/stable/model_selection.html#surprise.model_selection.validation.cross_validate
model_random_results = cross_validate(model_random, data, measures=['RMSE'], cv=5, verbose=True)

Evaluating RMSE of algorithm NormalPredictor on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.5233  1.5138  1.5141  1.5257  1.5241  1.5202  0.0052  
Fit time          0.53    0.73    0.57    0.70    0.70    0.64    0.08    
Test time         0.95    0.60    0.85    0.62    0.80    0.76    0.13    


# Model 2: User-Based Collaborative Filtering

In [23]:
# Create model using KNNBasic() class
# See the sim_options parameter to determine the user/item similarity calculation of the model
# http://surprise.readthedocs.io/en/stable/prediction_algorithms.html#similarity-measures-configuration
sim_options = {'user_based': True}
model_user = KNNBasic(sim_options=sim_options)

In [25]:
# Train using same cross validation code as above
model_user_results = cross_validate(model_user, data, measures=['RMSE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9856  0.9746  0.9751  0.9773  0.9830  0.9791  0.0044  
Fit time          1.23    1.35    1.14    1.01    1.15    1.18    0.11    
Test time         15.25   16.59   16.29   15.18   15.55   15.77   0.57    


# Model 3: Item-Based Collaborative Filtering

In [26]:
# Create model using KNNBasic() class
# Make sure you change the sim_options parameter from above
sim_options = {'user_based': False}
model_item = KNNBasic(sim_options=sim_options)

In [27]:
# Train using same cross validation code as above
model_item_results = cross_validate(model_item, data, measures=['RMSE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9729  0.9710  0.9800  0.9758  0.9685  0.9736  0.0040  
Fit time          8.36    12.16   1.96    1.93    2.21    5.32    4.21    
Test time         19.75   19.88   19.86   19.43   19.20   19.62   0.27    


<h1 style="color:red;">QUESTION 2: COLLABORATIVE FILTERING MODELS</h1>

**Compare the results from the user-user and item-item models. How do they compare to each other? How do they compare to our original "random" model? Can you provide any intuition as to why the results came out the way they did?**

#### 1. How do they compare to each other?

Mean RMSE user-user: 0.9788

Mean RMSE item-item: 0.9743

**Note**: item-item collaborative model has less average RMSE, as opposed to user-user. But the difference is in thousandths, and their average RMSE is approximately the same. Although test time of item-item collaborative model is comparably larger than user-user model (17.01 and 11.44, respectively)

#### 2. How do they compare to our original "random" model?

The average RMSE of random model is poorer than the user-user and item-item collaborative models (Mean RMSE random: 1.5167). But the time spent on testing data is much less. (Average test time = 0.58)
            
#### 3. Can you provide any intuition as to why the results came out the way they did?
The mean of RMSE of random model is higher because this algorithm uses no information about the data. That is why collaborative filtering models have better values of average RMSE as these models are personalized. We take into account individuals' features  and movies' details. The average RMSE user-user is less than average RMSE of item-item due to the certain reason. Collaborative filtering of users requires to compute inner products between rows (each user). The main constraint is that each user has rated a very small fraction of movies, and the raws are empty, in general. Therefore, it makes sense to focus on movies, which are most frequently rated (it means that they are popular). So we have a lot of ratings for these movies, and we can compute inner products between columns rather than rows. And in case of item-item model (columns), we have more collisions, as opposed to user-user collaborative model.

# Model 4: Matrix Factorization

In [36]:
# Create model using SVD() class
model_mf = SVD()

In [37]:
# Train using same cross validation code as above
model_mf_results = cross_validate(model_mf, data, measures=['RMSE'], cv=5, verbose=True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9396  0.9343  0.9282  0.9510  0.9305  0.9367  0.0081  
Fit time          25.27   45.18   47.84   42.22   29.07   37.91   9.03    
Test time         4.49    4.72    3.15    1.52    1.26    3.03    1.45    


<h1 style="color:red;">QUESTION 3: MATRIX FACTORIZATION MODEL</h1>

**The matrix factorization model is different from the collaborative filtering models. Briefly describe this difference. Also, compare the RMSE again. Does it improve? Can you offer any reasoning as to why that might be?**

#### 1. Difference between matrix factorization and collaborative filtering models:
Collaborative filtering models predict unknown user preferences based on existed rates and/or other users' behaviours. The main feature of these methods is the fact that users, who have similarly rated movies, have a tendency to similarly rate other movies.
Matrix factorization could reveal implicit factors related to user-item connection. The goal of matrix factorization is matrix expansion in product of two other matrices. The initial matrix is user-item. Missing ratings could be predicted by calculating squared error and gradient of it, and getting values of two matrices. So MF is the method of CF.  
#### 2. RMSE comparison:
The mean RMSE of CF is poorer than the average RMSE of MF. So it improved
#### 3. Can you offer any reasoning as to why that might be?
I think the difference is due to the fact that matrix factorization could find implicit patterns, as compared to CF models, which make predictions with clear correlations. 

# Precision and Recall @ `k`

We now want to compute the precision and recall for 2 values of `k`: 5 and 10. We have provided some code here to help you do that.

First, we define a function that takes in some predictions, a value of `k` and a threshold parameter. This code is adapted from [here](http://surprise.readthedocs.io/en/stable/FAQ.html?highlight=precision#how-to-compute-precision-k-and-recall-k).

In [38]:
def precision_recall_at_k(predictions, k=10, threshold=3.5):
    '''Return precision and recall at k metrics for each user.'''

    # First map the predictions to each user.
    user_est_true = dict()
    for uid, _, true_r, est, _ in predictions:
        current = user_est_true.get(uid, list())
        current.append((est, true_r))
        user_est_true[uid] = current

    precisions = dict()
    recalls = dict()
    for uid, user_ratings in user_est_true.items():

        # Sort user ratings by estimated value
        user_ratings.sort(key=lambda x: x[0], reverse=True)

        # Number of relevant items
        n_rel = sum((true_r >= threshold) for (_, true_r) in user_ratings)

        # Number of recommended items in top k
        n_rec_k = sum((est >= threshold) for (est, _) in user_ratings[:k])

        # Number of relevant and recommended items in top k
        n_rel_and_rec_k = sum(((true_r >= threshold) and (est >= threshold))
                              for (est, true_r) in user_ratings[:k])

        # Precision@K: Proportion of recommended items that are relevant
        precisions[uid] = n_rel_and_rec_k / n_rec_k if n_rec_k != 0 else 1

        # Recall@K: Proportion of relevant items that are recommended
        recalls[uid] = n_rel_and_rec_k / n_rel if n_rel != 0 else 1

    return precisions, recalls

Next, we compute the precision and recall at `k` = 5 and 10 for each of our 4 models. We use 5-fold cross validation again to average the results across the entire dataseat.

Please note that this will take some time to compute.

<h1 style="color:red;">QUESTION 4: PRECISION/RECALL</h1>

**Compute the precision and recall, for each of the 4 models, at `k` = 5 and 10. This is 2 x 2 x 4 = 16 numerical values. Do you note anything interesting about these values? Anything differerent from the RMSE values you computed above?**

Some code is required for this question.

In [40]:
# Use the function above to compute the 16 numerical values requested above
# See the test() function to get the predictions input to the function
# http://surprise.readthedocs.io/en/stable/algobase.html#surprise.prediction_algorithms.algo_base.AlgoBase.test
# Make list of k values
K = [5, 10]

# Make list of models
models = [model_random, model_user, model_item, model_mf]

# Create k-fold cross validation object
kf = KFold(n_splits=5)

for k in K:
    for model in models:
        print(f'>>> k={k}, model={model.__class__.__name__}')
        # Run folder and take average
        p = []
        r = []
        for trainset, testset in kf.split(data):
            model.fit(trainset)
            predictions = model.test(testset, verbose=False)
            precisions, recalls = precision_recall_at_k(predictions, k=k, threshold=3.5)

            # Precision and recall can then be averaged over all users
            p.append(sum(prec for prec in precisions.values()) / len(precisions))
            r.append(sum(rec for rec in recalls.values()) / len(recalls))
        
        print('>>> precision:', round(sum(p) / len(p), 3))
        print('>>> reccall  :', round(sum(r) / len(r), 3))
        print('\n')           

>>> k=5, model=NormalPredictor
>>> precision: 0.588
>>> reccall  : 0.342


>>> k=5, model=KNNBasic
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
>>> precision: 0.764
>>> reccall  : 0.455


>>> k=5, model=KNNBasic
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
>>> precision: 0.816
>>> reccall  : 0.39


>>> k=5, model=SVD
>>> precision: 0.78
>>> reccall  : 0.435


#### 1. Do you note anything interesting about these values? 
In random model increasing in two times of the number of clusters (k) leads to achieving high rates of precision and recall. What means, that this model predicts better with k = 10 instead of 5. Although, collaborative filtering models and matrix factorization method have other tendency. By increasing the number of clusters, the precisions of these models became less, as compared to the case, where k = 5. In comparison, recall rates behaved vice versa. When we set more clusters (10), recall values obtained higher points.
#### 2. Anything differerent from the RMSE values you computed above?
So RMSE is another type of metrics in comparison with precision and recall. In RMSE camputation, the best predictor was the matrix factorization method, as opposed to "precision/recall" metric, where in k = 5 user-based model achieved the highest precision (0.764) and recall (0.455). To compare RMSE with "precision/recall" metric (k=10), it is necessary to calculate RMSE for k =10.

#  Top-`n` Predictions

Finally, we can see what some of the actual movie ratings are for particular users, as outputs of our model.

Again, we define a helpful function.

In [41]:
def get_top_n(predictions, n=5):
    '''Return the top-N recommendation for each user from a set of predictions.

    Args:
        predictions(list of Prediction objects): The list of predictions, as
            returned by the test method of an algorithm.
        n(int): The number of recommendation to output for each user. Default
            is 10.

    Returns:
    A dict where keys are user (raw) ids and values are lists of tuples:
        [(raw item id, rating estimation), ...] of size n.
    '''

    # First map the predictions to each user.
    top_n = dict()
    for uid, iid, true_r, est, _ in predictions:
        current = top_n.get(uid, [])
        current.append((iid, est))
        top_n[uid] = current

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n

Then, we call this function on each of our models, first training on **all** the data we have available, then predicting on the remaining, missing data. We use `n`=5 here, but you can pick any reasonable value of `n` you would like.

This may take some time to compute, so be patient.

In [42]:
trainset = data.build_full_trainset()
testset = trainset.build_anti_testset()

Hint: Use [`**Dataset.build_full_trainset**`](http://surprise.readthedocs.io/en/stable/dataset.html#surprise.dataset.DatasetAutoFolds.build_full_trainset) to get the full trainset from the data. Then call [`**Trainset.build_anti_testset**`](http://surprise.readthedocs.io/en/stable/trainset.html#surprise.Trainset.build_anti_testset) to get the testset out. Finally, `fit` on the trainset, `test` on the testset, then pass that result to our `get_top_n` function.

<h1 style="color:red;">QUESTION 5: TOP N PREDICTIONS</h1>

**Do the top n predictions that you received make sense? What is the rating value (1-5) of these predictions? How could you use these predictions in the real-world if you were trying to build a generic content recommender system for a company?**

Some code is required for this question.

In [43]:
# Use the function and hints above to give the top-n predictions for a given user, for a reasonable value of n
for model in models:
    model.fit(trainset)
    predictions = model.test(testset)
    top_n = get_top_n(predictions, n=5)
    # Print the first one
    user = list(top_n.keys())[0]
    print(f'model: {model}, {user}: {top_n[user]}')

print('\n\nTop N computation successful!')

model: <surprise.prediction_algorithms.random_pred.NormalPredictor object at 0x7f6fa1cbd6a0>, 196: [('474', 5), ('387', 5), ('4', 5), ('100', 5), ('219', 5)]
Computing the msd similarity matrix...
Done computing similarity matrix.
model: <surprise.prediction_algorithms.knns.KNNBasic object at 0x7f6f849e3c18>, 196: [('1189', 5), ('1500', 5), ('814', 5), ('1536', 5), ('1599', 5)]
Computing the msd similarity matrix...
Done computing similarity matrix.
model: <surprise.prediction_algorithms.knns.KNNBasic object at 0x7f6fa1c84f28>, 196: [('1414', 4.666666666666667), ('1309', 4.5), ('1310', 4.5), ('1675', 4.333333333333333), ('1676', 4.3076923076923075)]
model: <surprise.prediction_algorithms.matrix_factorization.SVD object at 0x7f6f85718cc0>, 196: [('127', 4.737397575904313), ('64', 4.728400472602509), ('511', 4.628932304627258), ('272', 4.520357278067382), ('318', 4.514511059057053)]


Top N computation successful!


#### 1. Do the top n predictions that you received make sense? 
Yes, as for certain individual we have top 5 predictions of movies ratings calculated by 4 our methods. 
#### 2. What is the rating value (1-5) of these predictions?
The random model and user-based CF achieved 5 out of 5 for 5 movies. As compared to other two methods (item-based and MF), which have the ratings, which are close to 4.5.
#### 3. How could you use these predictions in the real-world if you were trying to build a generic content recommender system for a company?
I suppose that these predictions could be beneficial in case when we recommend movies with high rating to other people who have the same rates for other movies. It means that when we predict ratings of new data (movies) for new people, we could base on the previous received movies ratings (same users behaviours)

<hr>

Great job! Now, make sure you check out the **Conclusion** section of the [instruction manual](https://courses.edx.org/asset-v1:MITxPRO+DSx+2T2018+type@asset+block@4.1_instruction_manual.html) to wrap up this case study properly.